scripts/west_commands/build_helpers.py

```diff
@@ -25,21 +27,42 @@
 used; otherwise, it's "{}".'''.format(
     DEFAULT_BUILD_DIR)
 
+def _resolve_build_dir(fmt, cwd, **kwargs):
+    # Check if source_dir is below cwd first
+    source_dir = kwargs.get('source_dir')
+    if source_dir:
+        if escapes_directory(cwd, source_dir):
+            kwargs['source_dir'] = os.path.relpath(source_dir, cwd)
+        else:
+            # no meaningful relative path possible
+            kwargs['source_dir'] = ''
+    try:
+        return fmt.format(**kwargs)
+    except KeyError as k:
+        return None
 
-def find_build_dir(dir):
+def find_build_dir(dir, **kwargs):
     '''Heuristic for finding a build directory.
 
     If the given argument is truthy, it is returned. Otherwise, if
     the current working directory is a build directory, it is
-    returned. Otherwise, DEFAULT_BUILD_DIR is returned.'''
+    returned. Otherwise, if the build.dir-fmt configuration option is
+    set and can be resolved, it is returned. Otherwise, DEFAULT_BUILD_DIR
+    is returned.'''
     if dir:
         build_dir = dir
     else:
         cwd = os.getcwd()
         if is_zephyr_build(cwd):
             build_dir = cwd
         else:
-            build_dir = DEFAULT_BUILD_DIR
+            build_dir = config.get('build', 'dir-fmt', fallback=None)
```

Comment:
Shouldn't this be tested first (then build/, then cwd, as discussed elswhere)?

Reply:
Well, that really depends on what we choose to do. I actually used a an old comment of yours as a reference: https://github.com/zephyrproject-rtos/west/issues/124#issuecomment-448064044 but modified it slightly:

>1. -d is not given at the command line (highest precedence)
>2. $PWD is not a zephyr build directory
>3. $PWD/build (the default build directory) is not a zephyr build directory (lowest precedence)
>4. the directory implied by the west.command.build.build-dir-fmt option and command line arguments does not exist (i.e. if west build --board=my-board is given, build-dir-fmt = build/{board}, and build/my-board exists, use that)